test/integration/storageversion/storage_version_test.go

```diff
@@ -0,0 +1,114 @@
+	"k8s.io/kubernetes/test/integration/etcd"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+func TestStorageVersionCustomResource(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.StorageVersionAPI, true)()
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.APIServerIdentity, true)()
+	etcdConfig := framework.SharedEtcd()
+	server := kubeapiservertesting.StartTestServerOrDie(t, nil,
+		[]string{
+			"--runtime-config=internal.apiserver.k8s.io/v1alpha1=true",
+		},
+		etcdConfig)
+	crdClient := apiextensionsclientset.NewForConfigOrDie(server.ClientConfig)
+	crd := etcd.GetCustomResourceDefinitionData()[0]
+	etcd.CreateTestCRDs(t, crdClient, false, crd)
+	defer server.TearDownFn()
+	dynamicClient, err := dynamic.NewForConfig(server.ClientConfig)
+	if err != nil {
+		t.Fatalf("unexpected error creating dynamic client: %v", err)
+	}
+
+	kubeclient := kubernetes.NewForConfigOrDie(server.ClientConfig)
+	var apiserverID string
+	gr := crd.Spec.Group + "." + crd.Spec.Names.Plural
+	gvr := schema.GroupVersionResource{Group: crd.Spec.Group, Version: crd.Spec.Version, Resource: crd.Spec.Names.Plural}
+	// Send a request to make the server create handler and update storage version
+	_, err = dynamicClient.Resource(gvr).Namespace("default").List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		t.Fatalf("unexpected error when listing foos: %v", err)
+	}
+	createdCRD, err := crdClient.ApiextensionsV1().CustomResourceDefinitions().Get(context.TODO(), crd.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("failed to get crd: %v", err)
+	}
+
+	var storageVersion apiserverinternalv1alpha1.ServerStorageVersion
+	if err := wait.PollImmediate(100*time.Millisecond, 10*time.Second, func() (bool, error) {
+		sv, err := kubeclient.InternalV1alpha1().StorageVersions().Get(context.TODO(), gr, metav1.GetOptions{})
+		if err != nil {
+			return false, nil
+		}
+		ref := metav1.OwnerReference{
+			APIVersion: "apiextensions.k8s.io/v1",
+			Kind:       "CustomResourceDefinition",
+			Name:       createdCRD.Name,
+			UID:        createdCRD.UID,
+		}
+		// Owner reference gets assigned on create. Verify it first
+		if len(sv.OwnerReferences) == 0 || sv.OwnerReferences[0] != ref {
+			return false, fmt.Errorf("failed to set crd owner references, expect: %v, got: %v", ref, sv.OwnerReferences)
+		}
+		// The storageversion object gets created with empty status first, and then
+		// gets updated. Make sure we see the status.
+		if len(sv.Status.StorageVersions) != 1 {
+			return false, nil
+		}
+		storageVersion = sv.Status.StorageVersions[0]
+		return true, nil
+	}); err != nil {
+		t.Fatalf("failed to get storage version for custom resources: %v", err)
+	}
+	if !strings.HasPrefix(storageVersion.APIServerID, "kube-apiserver-") {
+		t.Fatalf("apiserver ID doesn't contain kube-apiserver- prefix, has: %v", apiserverID)
```

Comment:
i think apiserverID is "" at this point.

Reply:
It's populated by the apiserver-identity feature: https://github.com/kubernetes/kubernetes/blob/3146daf4334507f9bbc99b2ae333f9acb76fe3c3/staging/src/k8s.io/apiserver/pkg/server/config.go#L299